test/test_non_number_eval_tree_array.jl

```diff
@@ -0,0 +1,160 @@
+
+using DynamicExpressions
+using DynamicExpressions:
+    Node,
+    @extend_operators,
+    OperatorEnum,
+    get_constants,
+    append_number_constants!,
+    pop_number_constants,
+    count_number_constants
+
+# Max2Tensor (Tensor with a maximum of 3 dimensions) - struct that contains all three datatypes
+struct Max2Tensor{T}
+    dims::UInt8 # number of dimmentions
+    scalar::T
+    vector::Vector{T}
+    matrix::Matrix{T}
+    Max2Tensor{T}() where {T} = new(Int8(0), zero(T), T[], Array{T,2}(undef, 0, 0))
+    function Max2Tensor{T}(scalar::W) where {T,W<:Number}
+        return new(Int8(0), Base.convert(T, scalar), T[], Array{T,2}(undef, 0, 0))
+    end
+    function Max2Tensor{T}(vector::Vector{W}) where {T,W<:Number}
+        return new(Int8(1), zero(T), Vector{T}(vector), Array{T,2}(undef, 0, 0))
+    end
+    function Max2Tensor{T}(matrix::Matrix{W}) where {T,W<:Number}
+        return new(Int8(2), zero(T), T[], Matrix{T}(matrix))
+    end
+end
+
```

Comment:
Not sure if you are still planning to go in this direction (from the other struct you showed me) but just an idea: you could write this as:

```julia
julia> struct MultiTensor{T,N,A<:Tuple{Vararg{<:Any,N}}}
           dims::UInt8
           data::A
       
           function MultiTensor{T,N}() where {T,N}
               @assert N >= 1
               data = ntuple(i -> i == 1 ? zero(T) : zeros(T, ntuple(_ -> 0, Val(i-1))), Val(N))
               return new{T,N,typeof(data)}(zero(UInt8), data)
           end
       end
```

For example, for `N=3`, this would expand to:

```julia
function MultiTensor{T,N}() where {T,N}
    data = (zero(T), zeros(T, 0), zeros(T, 0, 0))
    return new{T,N,typeof(data)}(zero(UInt8), data)
end
```

and (I think) be type stable too, because `ntuple(..., Val(n))` will expand to exactly `n`-tuple if `n` is known to the compiler.

Then you can access the data with `val.data[1]` for scalars, `val.data[2]` for vectors, and so on.

(But I'm assuming you are going the other way with the fixed array?)